cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+    }
+
+    if revisions
+        .iter()
+        .any(|commit| commit.id() == workspace_command.repo().store().root_commit_id())
+    {
+        return Err(user_error("Cannot upload the virtual 'root()' commit"));
+    }
+
+    workspace_command.check_rewritable(revisions.iter().ids())?;
+
+    // If you have the changes main -> A -> B, and then run `jj gerrit upload -r B`,
+    // then that uploads both A and B. Thus, we need to ensure that A also
+    // has a Change-ID.
+    // We make an assumption here that all immutable commits already have a
+    // Change-ID.
+    let to_upload: Vec<Commit> = workspace_command
+        .attach_revset_evaluator(
+            // I'm unsure, but this *might* have significant performance
+            // implications. If so, we can change it to a maximum depth.
+            Rc::new(RevsetExpression::Difference(
+                // Unfortunately, DagRange{root: immutable_heads, heads: commits}
+                // doesn't work if you're, for example, working on top of an
+                // immutable commit that isn't in immutable_heads().
+                Rc::new(RevsetExpression::Ancestors {
+                    heads: RevsetExpression::commits(
+                        revisions.iter().ids().cloned().collect::<Vec<_>>(),
+                    ),
+                    generation: jj_lib::revset::GENERATION_RANGE_FULL,
+                    parents_range: jj_lib::revset::PARENTS_RANGE_FULL,
+                }),
+                workspace_command.env().immutable_expression().clone(),
+            )),
+        )
+        .evaluate_to_commits()?
+        .try_collect()?;
+
+    let mut tx = workspace_command.start_transaction();
+    let base_repo = tx.base_repo().clone();
+    let store = base_repo.store();
+
+    let old_heads = base_repo
+        .index()
+        .heads(&mut revisions.iter().ids())
+        .map_err(internal_error)?;
+
+    let git_settings = command.settings().git_settings()?;
+    let remote = calculate_push_remote(store, command.settings(), upload.remote.clone())?;
+    let remote_branch = calculate_push_ref(command.settings(), upload.remote_branch.clone())?;
+
+    // immediately error and reject any discardable commits, i.e. the
+    // the empty wcc
+    for commit in &to_upload {
+        if commit.is_discardable(tx.repo_mut())? {
+            return Err(user_error_with_hint(
+                format!(
+                    "Refusing to upload commit {} because it is an empty commit with no \
+                     description",
+                    short_commit_hash(commit.id())
+                ),
+                "Perhaps you squashed then ran upload? Maybe you meant to upload the parent \
+                 commit instead (eg. @-)",
+            ));
+        }
+    }
+
+    let mut old_to_new: IndexMap<CommitId, Commit> = IndexMap::new();
+    for commit_id in to_upload.iter().map(|c| c.id()).rev() {
+        let original_commit = store.get_commit(commit_id).unwrap();
```

Comment:
Avoid extracting just the id from the commit only in order to look up the commit. Do this instead:

```rust
for original_commit in to_upload {
```